Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-123.ca-central-1.compute.internal
 Git:  git version 2.9.5
Sha: 01892f8a71df819c5091e60734e0a53dd1d745d5
Containerized:False
Number of Error(s): 8
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
/home/centos/smoketest/PR-15278/HPCC-Platform/thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1798:28: error: no match for \xe2\x80\x98operator*\xe2\x80\x99 (operand type is \xe2\x80\x98std::vector<PartIO>\xe2\x80\x99)
/home/centos/smoketest/PR-15278/HPCC-Platform/thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp:1799:24: error: type \xe2\x80\x98<type error>\xe2\x80\x99 argument given to \xe2\x80\x98delete\xe2\x80\x99, expected pointer
make[2]: *** [thorlcr/activities/CMakeFiles/activityslaves_lcr.dir/build.make:401: thorlcr/activities/CMakeFiles/activityslaves_lcr.dir/keyedjoin/thkeyedjoinslave.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:17672: thorlcr/activities/CMakeFiles/activityslaves_lcr.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2021-09-15_11-44-00
time logs started
Prep time   :     8 sec (00:00:08)
Build time  :   376 sec (00:06:16)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   384 sec (00:06:24)
time logs end

[Build error:](https://gist.github.com/4099cd4a0ceef6148b15725623773c9b)
Build log [Build log](https://gist.github.com/HPCCSmoketest/4099cd4a0ceef6148b15725623773c9b#file-result-21-09-15-11-37-21-log)

--- a/system/jhtree/jhtree.cpp
+++ b/system/jhtree/jhtree.cpp
@@ -568,6 +568,11 @@ public:
         filter->describe(out);
     }
 
+    virtual void mergeStats(CRuntimeStatisticCollection & stats) const
+    {
+        if (keyCursor)
+            keyCursor->mergeStats(stats);
+    }
 };
 
 
@@ -1422,7 +1427,6 @@ CJHTreeNode *CKeyIndex::locateLastNode(KeyStatsCollector &stats)
         ::Release(last);
     }
 }
-
 
 void KeyStatsCollector::noteSeeks(unsigned lseeks, unsigned lscans, unsigned lwildseeks)
 {
@@ -2300,7 +2304,14 @@ public:
     virtual bool hasSpecialFileposition() const { return checkOpen().hasSpecialFileposition(); }
     virtual bool needsRowBuffer() const { return checkOpen().needsRowBuffer(); }
     virtual bool prewarmPage(offset_t offset, NodeType type) { return checkOpen().prewarmPage(offset, type); }
-
+    virtual void mergeStats(CRuntimeStatisticCollection & stats) const override
+    {
+        {
+            CriticalBlock b(c);
+            if (!realKey) return;
+        }
+        realKey->mergeStats(stats);
+    }
 };
 
 extern jhtree_decl IKeyIndex *createKeyIndex(const char *keyfile, unsigned crc, IFileIO &iFileIO, unsigned fileIdx, bool isTLK)

--- a/system/jhtree/jhtree.hpp
+++ b/system/jhtree/jhtree.hpp
@@ -79,6 +79,7 @@ interface jhtree_decl IKeyCursor : public IInterface
     virtual unsigned __int64 getCurrentRangeCount(unsigned groupSegCount, KeyStatsCollector &stats) = 0;
     virtual bool nextRange(unsigned groupSegCount) = 0;
     virtual const byte *queryKeyBuffer() const = 0;
+    virtual void mergeStats(CRuntimeStatisticCollection & stats) const = 0;
 };
 
 interface IKeyIndex;
@@ -116,6 +117,7 @@ interface jhtree_decl IKeyIndex : public IKeyIndexBase
     virtual bool hasSpecialFileposition() const = 0;
     virtual bool needsRowBuffer() const = 0;
     virtual bool prewarmPage(offset_t offset, NodeType type) = 0;
+    virtual void mergeStats(CRuntimeStatisticCollection & stats) const = 0;
 };
 
 interface IKeyArray : extends IInterface
@@ -286,6 +288,7 @@ interface IKeyManager : public IInterface, extends IIndexReadContext
     virtual unsigned getPartition() = 0;  // Use PARTITION() to retrieve partno, if possible, or zero to mean read all
 
     virtual unsigned numActiveKeys() const = 0;
+    virtual void mergeStats(CRuntimeStatisticCollection & stats) const = 0;
 };
 
 inline offset_t extractFpos(IKeyManager * manager)

--- a/system/jhtree/jhtree.ipp
+++ b/system/jhtree/jhtree.ipp
@@ -160,6 +160,8 @@ public:
     virtual CJHTreeNode * loadNode(CJHTreeNode * optNode, offset_t offset) = 0;
     CJHTreeNode *locateFirstNode(KeyStatsCollector &stats);
     CJHTreeNode *locateLastNode(KeyStatsCollector &stats);
+
+    virtual void mergeStats(CRuntimeStatisticCollection & stats) const override {}
 };
 
 class jhtree_decl CMemKeyIndex : public CKeyIndex
@@ -173,6 +175,7 @@ public:
     virtual const IFileIO *queryFileIO() const override { return nullptr; }
 // INodeLoader impl.
     virtual CJHTreeNode *loadNode(CJHTreeNode * optNode, offset_t offset);
+    virtual void mergeStats(CRuntimeStatisticCollection & stats) const override {}
 };
 
 class jhtree_decl CDiskKeyIndex : public CKeyIndex
@@ -188,6 +191,7 @@ public:
     virtual const IFileIO *queryFileIO() const override { return io; }
 // INodeLoader impl.
     virtual CJHTreeNode *loadNode(CJHTreeNode * optNode, offset_t offset);
+    virtual void mergeStats(CRuntimeStatisticCollection & stats) const override { ::mergeStats(stats, io); }
 };
 
 class jhtree_decl CKeyCursor : public CInterfaceOf<IKeyCursor>
@@ -250,6 +254,10 @@ protected:
     inline void endRange(unsigned segno)
     {
         filter->endRange(segno, keyBuffer);
+    }
+    virtual void mergeStats(CRuntimeStatisticCollection & stats) const override
+    {
+        key.mergeStats(stats);
     }
 };
 

--- a/thorlcr/activities/keyedjoin/thkeyedjoin.cpp
+++ b/thorlcr/activities/keyedjoin/thkeyedjoin.cpp
@@ -41,6 +41,7 @@ class CKeyedJoinMaster : public CMasterActivity
     bool remoteKeyedFetch = false;
     bool assumePrimary = false;
     unsigned totalIndexParts = 0;
+    std::vector<OwnedPtr<CThorStatsCollection>> fileStats;
 
     // CMap contains mappings and lists of parts for each slave
     class CMap
@@ -308,6 +309,11 @@ public:
             if (!isFileKey(indexFile))
                 throw MakeActivityException(this, TE_FileTypeMismatch, "Attempting to read flat file as an index: %s", indexFileName.get());
             IDistributedSuperFile *superIndex = indexFile->querySuperFile();
+            // One entry for each subfile (unless it is not a superfile => then add one entry for index data file stats)
+            unsigned numSuperIndexSubs = superIndex?superIndex->numSubFiles(true):1;
+            for (unsigned i=0; i<numSuperIndexSubs; i++)
+                fileStats.push_back(new CThorStatsCollection(indexReadStatistics));
+
             if (helper->diskAccessRequired())
             {
                 OwnedRoxieString fetchFilename(helper->getFileName());
@@ -351,6 +357,11 @@ public:
                                 remoteKeyedFetch = false;
                         }
                         dataMap.map(*this, dataFile, false, getOptBool("allLocalFetchParts"));
+                        IDistributedSuperFile *super = dataFile->querySuperFile();
+                        // One entry for each subfile (unless it is not a superfile => then have 1 entry for data file stats)
+                        unsigned numsubs = super?super->numSubFiles(true):1;
+                        for (unsigned i=0; i<numsubs; i++)
+                            fileStats.push_back(new CThorStatsCollection(diskReadRemoteStatistics));
                     }
                 }
             }
@@ -369,10 +380,8 @@ public:
                 indexFileDesc.setown(indexFile->getFileDescriptor());
 
                 unsigned superIndexWidth = 0;
-                unsigned numSuperIndexSubs = 0;
                 if (superIndex)
                 {
-                    numSuperIndexSubs = superIndex->numSubFiles(true);
                     bool first=true;
                     // consistency check
                     Owned<IDistributedFileIterator> iter = superIndex->getSubFileIterator(true);
@@ -458,10 +467,7 @@ public:
                     initMb.append(keyHasTlk);
                     if (keyHasTlk)
                     {
-                        if (numSuperIndexSubs)
-                            initMb.append(numSuperIndexSubs);
-                        else
-                            initMb.append((unsigned)1);
+                        initMb.append(numSuperIndexSubs);
 
                         Owned<IDistributedFileIterator> iter;
                         IDistributedFile *f;
@@ -551,6 +557,23 @@ public:
                     dataMap.serializePartMap(dst);
             }
         }
+    }
+    virtual void deserializeStats(unsigned node, MemoryBuffer &mb) override
+    {
+        CMasterActivity::deserializeStats(node, mb);
+        for (auto &stats: fileStats)
+            stats->deserialize(node, mb);
+    }
+    virtual void done() override
+    {
+        unsigned numSubFiles = fileStats.size();
+        for (unsigned i=0; i<numSubFiles; i++)
+        {
+            IDistributedFile *file = queryReadFile(i);
+            if (file) //publish the number of disk reads for indexes and disk fetches.
+                file->addAttrValue("@numDiskReads", fileStats[i]->getStatisticSum(StNumDiskReads));
+        }
+        CMasterActivity::done();
     }
 };
 

--- a/thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
+++ b/thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp
@@ -56,6 +56,7 @@
 #include <algorithm>
 #include <typeinfo>
 #include <type_traits>
+#include <memory>
 
 static const unsigned defaultKeyLookupQueuedBatchSize = 1000;
 static const unsigned defaultKeyLookupFetchQueuedBatchSize = 1000;
@@ -974,14 +975,16 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
         std::vector<CThorExpandingRowArray *> batchArrays;
         std::vector<unsigned> parts; // part numbers this handler is dealing with
         std::vector<unsigned> partNumMap; // map of part # -> parts index
+        std::vector<unsigned> subFileNum;
         unsigned myParts = 0;
         unsigned nextQueue = 0;
         bool stopped = false;
         unsigned lookupQueuedBatchSize = 1000;
         rowcount_t total = 0;
         CLimiter *limiter = nullptr;
-        IArrayOf<IPartDescriptor> *allParts = nullptr; // only used for tracing purposes, set by key or fetch derived handlers
+        IArrayOf<IPartDescriptor> *allParts = nullptr;
         unsigned batchProcessLimit = 0xffffffff;
+        bool isSuper = false;
 
         inline MemoryBuffer &doUncompress(MemoryBuffer &tgt, MemoryBuffer &src)
         {
@@ -1051,6 +1054,18 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
             nextQueue = 0;
             totalQueued = 0;
             state = ts_initial;
+            ISuperFileDescriptor * superFDesc = allParts->item(0).queryOwner().querySuperFileDescriptor();
+            if (superFDesc)
+            {
+                unsigned subfile, lnum;
+                for (size_t i=0; i<parts.size(); i++)
+                {
+                    unsigned partNo = parts[i] & partMask;
+                    verifyex(superFDesc->mapSubPart(partNo, subfile, lnum));
+                    subFileNum.push_back(subfile);
+                }
+                isSuper=true;
+            }
         }
         void join()
         {
@@ -1224,6 +1239,7 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
             if (limiter)
                 limiter->dec(); // unblocks any requests to start lookup threads
         }
+        virtual void getSubFileStats(std::vector<OwnedPtr<CRuntimeStatisticCollection>> & subFileStats) = 0;
     };
 
     class CKeyLookupLocalBase : public CLookupHandler
@@ -1332,23 +1348,30 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
     {
         typedef CKeyLookupLocalBase PARENT;
 
-        std::vector<IKeyManager *> keyManagers;
+        //Need ptr to std::vector as std::atomic's are not constructable(doesn't have copy constructor)
+        std::unique_ptr<std::vector<std::atomic<IKeyManager *>>> keyManagers;
     public:
         CKeyLookupLocalHandler(CKeyedJoinSlave &_activity) : CKeyLookupLocalBase(_activity)
         {
         }
         ~CKeyLookupLocalHandler()
         {
-            for (auto &km : keyManagers)
+            for (auto &km : *keyManagers)
             {
                 if (km)
-                    km->Release();
+                    (*km).Release();
             }
+        }
+        virtual void init() override
+        {
+            PARENT::init();
+            keyManagers.reset(new std::vector<std::atomic<IKeyManager *>>(parts.size()));
+            for (auto & k: *keyManagers)
+                k = nullptr;
         }
         virtual void addPartNum(unsigned partNum) override
         {
             PARENT::addPartNum(partNum);
-            keyManagers.push_back(nullptr);
             translators.push_back(nullptr);
         }
         virtual void process(CThorExpandingRowArray &processing, unsigned selected) override
@@ -1356,7 +1379,7 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
             unsigned partCopy = parts[selected];
             unsigned partNo = partCopy & partMask;
             unsigned copy = partCopy >> partBits;
-            IKeyManager *&keyManager = keyManagers[selected];
+            std::atomic<IKeyManager *> & keyManager = (*keyManagers)[selected];
             if (!keyManager) // delayed until actually needed
             {
                 keyManager = activity.createPartKeyManager(partNo, copy);
@@ -1365,17 +1388,38 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
             }
             processRows(processing, partNo, keyManager);
         }
+        virtual void getSubFileStats(std::vector<OwnedPtr<CRuntimeStatisticCollection>> & subFileStats) override
+        {
+            for (size_t i=0; i<parts.size(); i++)
+            {
+                IKeyManager *keyManager = (*keyManagers)[i];
+                if (!keyManager)
+                    continue;
+                if (isSuper)
+                {
+                    unsigned subfile = subFileNum[i];
+                    keyManager->mergeStats(*subFileStats[subfile]);
+                }
+                else
+                    keyManager->mergeStats(*subFileStats[0]);
+            }
+        }
     };
     class CKeyLookupMergeHandler : public CKeyLookupLocalBase
     {
         typedef CKeyLookupLocalBase PARENT;
 
-        Owned<IKeyManager> keyManager;
+        std::atomic<IKeyManager *> keyManager{nullptr};
     public:
         CKeyLookupMergeHandler(CKeyedJoinSlave &_activity) : CKeyLookupLocalBase(_activity)
         {
             limiter = &activity.lookupThreadLimiter;
             translators.push_back(nullptr);
+        }
+        ~CKeyLookupMergeHandler()
+        {
+            if (keyManager)
+                (*keyManager).Release();
         }
         virtual void process(CThorExpandingRowArray &processing, unsigned __unused) override
         {
@@ -1389,11 +1433,28 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
                     Owned<IKeyIndex> keyIndex = activity.createPartKeyIndex(partNo, copy, false);
                     partKeySet->addIndex(keyIndex.getClear());
                 }
-                keyManager.setown(createKeyMerger(helper->queryIndexRecordSize()->queryRecordAccessor(true), partKeySet, 0, nullptr, helper->hasNewSegmentMonitors(), false));
+                keyManager = createKeyMerger(helper->queryIndexRecordSize()->queryRecordAccessor(true), partKeySet, 0, nullptr, helper->hasNewSegmentMonitors(), false);
                 setupTranslation(0, 0, *keyManager);
             }
             processRows(processing, 0, keyManager);
         }
+        virtual void getSubFileStats(std::vector<OwnedPtr<CRuntimeStatisticCollection>> & subFileStats) override
+        {
+            if (keyManager)
+            {
+                for (size_t i=0; i<parts.size(); i++)
+                {
+                    if (isSuper)
+                    {
+                        unsigned subfile = subFileNum[i];
+                        (*keyManager).mergeStats(*subFileStats[subfile]);
+                    }
+                    else
+                        (*keyManager).mergeStats(*subFileStats[0]);
+                }
+            }
+        }
+
     };
     class CRemoteLookupHandler : public CLookupHandler
     {
@@ -1477,6 +1538,10 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
             handles.resize(parts.size());
             for (auto &h: handles)
                 h = 0;
+        }
+        virtual void getSubFileStats(std::vector<OwnedPtr<CRuntimeStatisticCollection>> & subFileStats) override
+        {
+            /* Note: currently, stats from remote file not tracked */
         }
     };
     class CKeyLookupRemoteHandler : public CRemoteLookupHandler
@@ -1705,12 +1770,16 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
         Owned<IEngineRowAllocator> fetchDiskAllocator;
         CThorContiguousRowBuffer prefetchSource;
         std::vector<PartIO> partIOs;
+        std::unique_ptr<std::vector<std::atomic_bool>> partIOCreated;
 
         inline const PartIO &queryFetchPartIO(unsigned selected, unsigned partNo, unsigned copy, bool compressed, bool encrypted)
         {
             PartIO &partIO = partIOs[selected];
             if (!partIO.stream)
+            {
                 partIO = activity.getFetchPartIO(partNo, copy, compressed, encrypted);
+                (*partIOCreated)[selected] = true;
+            }
             return partIO;
         }
     public:
@@ -1728,6 +1797,9 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
             encrypted = activity.allDataParts.item(0).queryOwner().queryProperties().getPropBool("@encrypted");
             compressed = isCompressed(activity.allDataParts.item(0).queryOwner().queryProperties());
             partIOs.resize(parts.size());
+            partIOCreated.reset(new std::vector<std::atomic_bool>(parts.size()));
+            for (auto & p: *partIOCreated)
+                p = false;
         }
         virtual void process(CThorExpandingRowArray &processing, unsigned selected) override
         {
@@ -1789,6 +1861,23 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
                 joinGroup->decPending(); // Every queued lookup row triggered an inc., this is the corresponding dec.
 
                 diskSeeks++;
+            }
+        }
+        virtual void getSubFileStats(std::vector<OwnedPtr<CRuntimeStatisticCollection>> & subFileStats) override
+        {
+            for (size_t i=0; i<parts.size(); i++)
+            {
+                if ((*partIOCreated)[i])
+                {
+                    PartIO & partIO = partIOs[i];
+                    if (isSuper)
+                    {
+                        unsigned subfile = subFileNum[i];
+                        mergeStats(*subFileStats[subfile], partIO.iFileIO);
+                    }
+                    else
+                        mergeStats(*subFileStats[0], partIO.iFileIO);
+                }
             }
         }
     };
@@ -2092,6 +2181,11 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
                     lookupHandler->end();
             }
         }
+        void getSubFileStats(std::vector<OwnedPtr<CRuntimeStatisticCollection>> & subFileStats)
+        {
+            ForEachItemIn(h, handlers)
+                handlers.item(h)->getSubFileStats(subFileStats);
+        }
     };
 
     const unsigned startSpillAmountPercentage = 10;   // Initial percentage to spill, will grow if memory pressure keeps calling callback
@@ -2131,7 +2225,6 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
     std::vector<FPosTableEntry> globalFPosToSlaveMap; // maps fpos->part
     std::vector<unsigned> indexPartToSlaveMap;
     std::vector<unsigned> dataPartToSlaveMap;
-
 
     unsigned atMost = 0, keepLimit = 0;
     unsigned abortLimit = 0;
@@ -2187,6 +2280,8 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
 
     CriticalSection fetchFileCrit;
     std::vector<PartIO> openFetchParts;
+    std::vector<OwnedPtr<CRuntimeStatisticCollection>> dataFileStats;
+    std::vector<OwnedPtr<CRuntimeStatisticCollection>> indexFileStats;
 
     PartIO getFetchPartIO(unsigned partNo, unsigned copy, bool compressed, bool encrypted)
     {
@@ -2333,8 +2428,8 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
 
         if (delayed)
         {
-            Owned<IFileIO> lazyFileIO = queryThor().queryFileCache().lookupIFileIO(*this, indexName, filePart);
-            Owned<IDelayedFile> delayedFile = createDelayedFile(lazyFileIO);
+            Owned<IFileIO> lazyIFileIO = queryThor().queryFileCache().lookupIFileIO(*this, indexName, filePart, nullptr, indexReadStatistics);
+            Owned<IDelayedFile> delayedFile = createDelayedFile(lazyIFileIO);
             return createKeyIndex(filename, crc, *delayedFile, (unsigned) -1, false);
         }
         else
@@ -2343,7 +2438,7 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor, implem
              * But that's okay, because we are only here on demand.
              * The underlying IFileIO can later be closed by fhe file caching mechanism.
              */
-            Owned<IFileIO> lazyIFileIO = queryThor().queryFileCache().lookupIFileIO(*this, indexName, filePart);
+            Owned<IFileIO> lazyIFileIO = queryThor().queryFileCache().lookupIFileIO(*this, indexName, filePart, nullptr, indexReadStatistics);
             return createKeyIndex(filename, crc, *lazyIFileIO, (unsigned) -1, false);
         }
     }
@@ -2931,6 +3026,8 @@ public:
             dataPartToSlaveMap.clear();
             tlkKeyManagers.kill();
             partitionKey = false;
+            dataFileStats.clear();
+            indexFileStats.clear();
         }
         // decode data from master. NB: can be resent and differ if in global loop
         data.read(indexName);
@@ -3003,8 +3100,13 @@ public:
                 else if (container.queryLocalData())
                     totalIndexParts = numIndexParts; // will be same unless local data only
             }
+            ISuperFileDescriptor *superFdesc = numIndexParts?allIndexParts.item(0).queryOwner().querySuperFileDescriptor():nullptr;
+            // One entry in indexFileStats vector for each index file's subfile
+            // Unless it is not a superfile => then the indexFileStats[0] will be used for the index file's stats
+            unsigned numIndexSubFiles = superFdesc?superFdesc->querySubFiles():1;
+            for (unsigned n=0; n<numIndexSubFiles; ++n)
+                indexFileStats.push_back(new CRuntimeStatisticCollection(indexReadStatistics));
 
-            ISuperFileDescriptor *superFdesc = numIndexParts ? allIndexParts.item(0).queryOwner().querySuperFileDescriptor() : nullptr;
             setupLookupHandlers(keyLookupHandlers, totalIndexParts, superFdesc, localIndexParts, indexPartToSlaveMap, localKey, forceRemoteKeyedLookup ? ht_remotekeylookup : ht_localkeylookup, ht_remotekeylookup);
             data.read(totalDataParts);
             if (totalDataParts)
@@ -3040,6 +3142,11 @@ public:
                     e->index = f;
                 }
                 std::sort(globalFPosToSlaveMap.begin(), globalFPosToSlaveMap.end(), [](const FPosTableEntry &a, const FPosTableEntry &b) { return a.base < b.base; });
+                // One entry in dataFileStats vector for each data file's subfile
+                // Unless it is not a superfile => then the dataFileStats[0] will be used for the data file's stats
+                unsigned numDataSubFiles = superFdesc?superFdesc->querySubFiles():1;
+                for (unsigned n=0; n<numDataSubFiles; ++n)
+                    dataFileStats.push_back(new CRuntimeStatisticCollection(diskReadRemoteStatistics));
 #ifdef _DEBUG
                 for (unsigned c=0; c<totalDataParts; c++)
                 {
@@ -3672,11 +3779,23 @@ public:
     {
         return container.queryId();
     }
+    virtual void serializeStats(MemoryBuffer &mb) override
+    {
+        keyLookupHandlers.getSubFileStats(indexFileStats);
+        if (dataFileStats.size()>0)
+            fetchLookupHandlers.getSubFileStats(dataFileStats);
+
+        PARENT::serializeStats(mb);
+        for (auto &stats: indexFileStats)
+            stats->serialize(mb);
+        for (auto &stats: dataFileStats)
+            stats->serialize(mb);
+    }
 };
 
 
-CActivityBase *createKeyedJoinSlave(CGraphElementBase *container) 
-{ 
+CActivityBase *createKeyedJoinSlave(CGraphElementBase *container)
+{
     if (container->getOptBool("legacykj"))
         return LegacyKJ::createKeyedJoinSlave(container);
     return new CKeyedJoinSlave(container);

--- a/thorlcr/thorutil/thormisc.cpp
+++ b/thorlcr/thorutil/thormisc.cpp
@@ -75,7 +75,8 @@ const StatisticsMapping spillStatistics({StTimeSpillElapsed, StTimeSortElapsed, 
 const StatisticsMapping basicActivityStatistics({StTimeLocalExecute, StTimeBlocked});
 const StatisticsMapping groupActivityStatistics({StNumGroups, StNumGroupMax}, basicActivityStatistics);
 const StatisticsMapping hashJoinActivityStatistics({StNumLeftRows, StNumRightRows}, basicActivityStatistics);
-const StatisticsMapping indexReadActivityStatistics({StNumRowsProcessed, StNumIndexSeeks, StNumIndexScans, StNumPostFiltered, StNumIndexWildSeeks}, diskReadRemoteStatistics, basicActivityStatistics);
+const StatisticsMapping indexReadStatistics({StNumIndexSeeks, StNumIndexScans, StNumPostFiltered, StNumIndexWildSeeks});
+const StatisticsMapping indexReadActivityStatistics({StNumRowsProcessed}, diskReadRemoteStatistics, basicActivityStatistics, indexReadStatistics);
 const StatisticsMapping indexWriteActivityStatistics({StPerReplicated}, basicActivityStatistics, diskWriteRemoteStatistics);
 const StatisticsMapping keyedJoinActivityStatistics({ StNumIndexSeeks, StNumIndexScans, StNumIndexAccepted, StNumPostFiltered, StNumPreFiltered, StNumDiskSeeks, StNumDiskAccepted, StNumDiskRejected, StNumIndexWildSeeks}, basicActivityStatistics);
 const StatisticsMapping loopActivityStatistics({StNumIterations}, basicActivityStatistics);

--- a/thorlcr/thorutil/thormisc.hpp
+++ b/thorlcr/thorutil/thormisc.hpp
@@ -137,7 +137,7 @@ extern graph_decl const StatisticsMapping diskWriteActivityStatistics;
 extern graph_decl const StatisticsMapping sortActivityStatistics;
 
 extern graph_decl const StatisticsMapping graphStatistics;
-
+extern graph_decl const StatisticsMapping indexReadStatistics;
 
 class BooleanOnOff
 {

